### PR TITLE
Compare identifier using #new_edition_of?

### LINKED ISSUE #49: Pubid algebra: is new edition of
related #47 

> is new edition of (newer edition year)

```ruby
Identifier.parse("ISO 22610:2018").new_edition_of?("ISO 22610:2006") => true
```

### PR DESCRIPTION
**Pull request by mico**
closes #49

In this implementation I only consider year to compare editions, numbers are ignored.
Need clarification on edition numbers here https://github.com/metanorma/pubid-iso/issues/274

### CODE REVIEW

**Comment by andrew2net** [comment]
As I understand this PR makes pubid-core ignore edition numbers because ISO IDs don't have editions.
I want to draw attention to fact that some standards have edition numbers in their IDs. For example IHO standards look like `IHO S-100 1.0.0` where `1.0.0` is an edition number.
Shouldn't we implement ISO specific behavior in pubid-iso?

**Comment by mico** [comment]
> As I understand this PR makes pubid-core ignore edition numbers because ISO IDs don't have editions. I want to draw attention to fact that some standards have edition numbers in their IDs. For example IHO standards look like `IHO S-100 1.0.0` where `1.0.0` is an edition number. Shouldn't we implement ISO specific behavior in pubid-iso?

This PR ignore edition numbers when comparing identifiers using `#new_edition_of?` it doesn't affect anything else.
The reason — I still don't understand how we should compare edition numbers.
ISO have edition numbers, it was mentioned in https://github.com/metanorma/pubid-iso/issues/274

But as I understand different publishers might have different approach to edition number, so we can implement separate implementation (or extension) of `#new_edition_of?` for each specific `pubid-*` gem where edition numbers need to be considered for comparison.

File: README.adoc [hunk 1]
```diff
@@ -29,3 +28,20 @@ pubid = Identifier.parse("ISO 1:1999")
...
+
+pubid_first = Identifier.parse("ISO 1:1999")
+pubid_second = Identifier.parse("ISO 1")
+
+pubid_first.new_edition_of?(pubid_second)
+=> false
+
+pubid_second = Identifier.parse("ISO 1:1998")
+pubid_first.new_edition_of?(pubid_second)
+=> false
```
Comment (ronaldtse) [inline README.adoc]: This should add a reversed check:

```ruby
pubid_second.new_edition_of?(pubid_first)
=> true
```

File: lib/pubid/core/identifier/base.rb [hunk 2]
```diff
@@ -167,6 +167,19 @@ def resolve_stage(stage)
...
+      # @param other [Pubid::Core::Identifier::Base] pubid identifier to compare with
+      # @return [Boolean] true if another identifier is newer edition
+      def new_edition_of?(other)
+        raise Errors::AnotherDocumentError, "cannot compare edition with #{other}" unless exclude(:year) == other.exclude(:year)
+
+        return false if year.nil?
+
+        return true if other.year.nil?
+
+        year < other.year
```
Comment (ronaldtse) [inline lib/pubid/core/identifier/base.rb]: You need to also check the case where if they have edition numbers, some may not have year but have edition.

**Review by ronaldtse — CHANGES_REQUESTED** [review]

File: lib/pubid/core/identifier/base.rb (hunk 2 above)
Comment (mico) [reply lib/pubid/core/identifier/base.rb]: > You need to also check the case where if they have edition numbers, some may not have year but have edition.

@ronaldtse I have question here about that: https://github.com/metanorma/pubid-iso/issues/274
What if we have identifier year 1999, edition number 2 and another identifier year 2000 and edition number 1, is it possible? Does edition number depends on year or they are independed? Is it the same for all publishers? (ISO, IEEE, NIST, ITU, etc)

Comment (ronaldtse) [reply lib/pubid/core/identifier/base.rb]: It is possible, what that means is that the new edition 1 is a different document altogether from the old document. 

This happens when a standard document (eg 10001:2000 edition 3) is re-numbered to another number (eg 11001:2001 edition 4), then a new document adopts the old number (eg 10001:2002 edition 1). 

So this means these two documents are "incomparable". As @andrew2net mentioned, we need to distinguish cases where the documents cannot be compared. Either use NIL or throw an error.

**Review by andrew2net — APPROVED** [review]

**Pull request merged**